### DISCUSSION TITLE
JPA 엔티티 동등성 비교

### DISCUSSION THREAD

**Discussion by Kim0914** [woowacourse-teams/2023-pium#163] [뜨거운 토론]
### Case 1, 자바에서 생성해주는 equals and hashcode 메서드로 비교 (SubClass 허용 X)

결과: 실패 -> getClass() 비교에서 실패

### Case 2, 자바에서 생성해주는 equals and hashcode 메서드로 비교 (SubClass 허용 O)

결과: 성공 -> `(o instanceOf Member member)`로 검증한 후 getId()를 해서 비교

### Case 3, 롬복에서 생성해주는 equals and hashcode 메서드로 비교

결과: 성공 -> 동일하게 getId()로 비교함

```java
@Override
    public final boolean equals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Class<?> oEffectiveClass = o instanceof HibernateProxy ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass() : o.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass() : this.getClass();
        if (thisEffectiveClass != oEffectiveClass) return false;
        DictionaryPlant that = (DictionaryPlant) o;
        return getId() != null && Objects.equals(getId(), that.getId());
    }
```
결론: 인텔리제이에서 제공해주는 프록시까지 고려한 equals and hashcode를 사용한다 !

**Comment by Choi-JJunho** [comment]
해당 내용 코드컨벤션에 추가완료 하였습니다.
- https://github.com/woowacourse-teams/2023-pium/wiki/%5BBE%5D-%EC%BD%94%EB%93%9C-%EC%BB%A8%EB%B2%A4%EC%85%98
